src/electron/go_vpn_tunnel.ts

```diff
@@ -0,0 +1,316 @@
+// Copyright 2021 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {execFile} from 'child_process';
+import {powerMonitor} from 'electron';
+import {platform} from 'os';
+import {promisify} from 'util';
+
+import {TunnelStatus} from '../www/app/tunnel';
+import * as errors from '../www/model/errors';
+import {ShadowsocksConfig} from '../www/model/shadowsocks';
+
+import {RoutingDaemon} from './routing_service';
+import {ChildProcessHelper} from './process';
+import {pathToEmbeddedBinary} from './util';
+import {VpnTunnel} from './vpn_tunnel';
+
+const isLinux = platform() === 'linux';
+const isWindows = platform() === 'win32';
+
+const TUN2SOCKS_TAP_DEVICE_NAME = isLinux ? 'outline-tun0' : 'outline-tap0';
+const TUN2SOCKS_TAP_DEVICE_IP = '10.0.85.2';
+const TUN2SOCKS_VIRTUAL_ROUTER_IP = '10.0.85.1';
+const TUN2SOCKS_VIRTUAL_ROUTER_NETMASK = '255.255.255.0';
+
+// Cloudflare and Quad9 resolvers.
+const DNS_RESOLVERS = ['1.1.1.1', '9.9.9.9'];
+
+// Establishes a full-system VPN with the help of Outline's routing daemon and child process
+// outline-go-tun2socks. The routing service modifies the routing table so that the TAP device
+// receives all device traffic. outline-go-tun2socks process TCP and UDP traffic from the TAP
+// device and relays it to a Shadowsocks proxy server.
+//
+// |TAP| <-> |outline-go-tun2socks| <-> |Shadowsocks proxy|
+//
+// In addition to the basic lifecycle of the helper processes, this class restarts tun2socks
+// on network changes if necessary.
+// Follows the Mediator pattern in that none of the "helpers" know anything
+// about the others.
+export class GoVpnTunnel implements VpnTunnel {
+  private readonly routing: RoutingDaemon;
+  private readonly tun2socks: GoTun2socks;
+
+  // Extracted out to an instance variable because in certain situations, notably a change in UDP
+  // support, we need to stop and restart tun2socks *without notifying the client* and this allows
+  // us swap the listener in and out.
+  private tun2socksExitListener?: () => void | undefined;
+
+  // See #resumeListener.
+  private terminated = false;
+
+  private isUdpEnabled = false;
+
+  private readonly onAllHelpersStopped: Promise<void>;
+
+  private reconnectingListener?: () => void;
+
+  private reconnectedListener?: () => void;
+
+  constructor(private config: ShadowsocksConfig, private isAutoConnect: boolean) {
```

Comment:
What is the `isAutoConnect` option?

Reply:
It tells the tunnel whether to check connectivity; the routing service needs it to enforce a kill switch.
Since the routing service is now injected, I refactored this property into an argument to `connect`.